compiler/util-klib-abi/src/org/jetbrains/kotlin/library/abi/parser/KlibParsingCursorExtensions.kt

```diff
@@ -0,0 +1,526 @@
+
+/**
+ * Used to check if declarations after a property are getter / setter methods which should be
+ * attached to that property.
+ */
+private fun Cursor.hasPropertyAccessor(type: GetterOrSetter): Boolean {
+    val subCursor = copy()
+    subCursor.parseAbiModality()
+    subCursor.parseFunctionModifiers()
+    subCursor.parseFunctionKind() ?: return false // if it's not a function it's not a getter/setter
+    val mightHaveTypeParams = subCursor.parseGetterOrSetterName(peek = true) == null
+    if (mightHaveTypeParams) {
+        subCursor.parseTypeParams()
+    }
+    subCursor.parseContextAndReceiverParams()
+    return when (type) {
+        GetterOrSetter.GETTER -> subCursor.parseGetterName() != null
+        GetterOrSetter.SETTER -> subCursor.parseSetterName() != null
+    }
+}
+
+private fun Cursor.subCursor(peek: Boolean) =
+    if (peek) {
+        copy()
+    } else {
+        this
+    }
+
+private fun Cursor.parseTypeParamsString(peek: Boolean = false): String? {
+    if (parseSymbol(getterOrSetterSignalRegex, peek = true) != null) {
+        return null
+    }
+    val cursor = subCursor(peek)
+    val result = StringBuilder()
+    cursor.parseSymbol(openAngleBracketRegex)?.let { result.append(it) } ?: return null
+    var openBracketCount = 1
+    while (openBracketCount > 0) {
+        val nextSymbol =
+            cursor.parseSymbol(anyCharRegex, skipInlineWhitespace = false).also {
+                result.append(it)
+            }
+        when (nextSymbol) {
+            "<" -> openBracketCount++
+            ">" -> openBracketCount--
+        }
+    }
+    cursor.skipInlineWhitespace()
+    return result.toString()
+}
+
+private fun Cursor.parseAbiModalityString(peek: Boolean = false) =
+    parseSymbol(abiModalityRegex, peek)?.uppercase()
+
+private fun Cursor.parsePropertyKindString(peek: Boolean = false) =
+    parseSymbol(propertyKindRegex, peek)?.uppercase()?.replace(" ", "_")
+
+private fun Cursor.parseClassKindString(peek: Boolean = false) =
+    parseSymbol(classKindRegex, peek)?.uppercase()?.replace(" ", "_")
+
+private enum class GetterOrSetter() {
+    GETTER,
+    SETTER,
+}
+
+private val packageNameRegex = Regex("^[^/]*")
+private val constructorNameRegex = Regex("^constructor\\s<init>")
+private val closeCurlyBraceRegex = Regex("^}")
+private val uniqueNameMarkerRegex = Regex("^Library unique name: ")
+private val uniqueNameRegex = Regex("[a-zA-Z\\-\\.:]+")
+private val commentMarkerRegex = Regex("^\\/\\/")
+private val anyCharRegex = Regex(".")
+private val closeSquareBracketRegex = Regex("^\\]")
+private val openSquareBracketRegex = Regex("^\\[")
+private val defaultArgSymbolRegex = Regex("^=(\\s)?\\.\\.\\.")
+private val varargSymbolRegex = Regex("^\\.\\.\\.")
+private val openParenRegex = Regex("^\\(")
+private val closeParenRegex = Regex("^\\)")
+private val reifiedRegex = Regex("reified")
+private val contextRegex = Regex("^context")
+private val colonRegex = Regex("^:")
```

Comment:
Don't get me wrong, I'm not suggesting to win a "1 billion klib dumps challenge" or something, but a few tweaks could be applied to the current implementation to keep it relatively simple, yet make it a bit faster.

Some of the regexes below don't have to be regexes at all: it's either a literal string, or set of literal strings; by introducing a `parseSymbol` overload capable of handling those (and not creating substring to check the current line), the performance could be improved.